### PR TITLE
[WIP] Proxy between Ethereum network and ours

### PR DESCRIPTION
**Pull request by adambabik**
We want to enable LES support but as LES peers are in Ethereum network, we must advertise them within ours.

This is an experiment. Before merging this PR, we should wait for rendezvous peer discovery protocol. We may end up with running rendezvous and one Discovery V5 and proxy part will be implemented between rendezvous and Discovery V5 instead of Discovery V5 and Discovery V5.

### CODE REVIEW

File: node/status_node.go [hunk 1]
```diff
@@ -107,7 +114,30 @@ func (n *StatusNode) startWithDB(config *params.NodeConfig, db *leveldb.DB, serv
 	}
 
 	if n.discoveryEnabled() {
-		return n.startDiscovery()
+		// Check if LES is enabled before starting Discovery.
+		// If so, add LES topics to required topics.
+		if config.LightEthConfig != nil && config.LightEthConfig.Enabled {
```
Comment (dshulyak) [inline node/status_node.go:119]: btw les code by itself uses discovery v5 server. implementation is somewhat similar to peer pool, but it never stops.

File: node/status_node.go [hunk 2]
```diff
@@ -107,7 +114,30 @@ func (n *StatusNode) startWithDB(config *params.NodeConfig, db *leveldb.DB, serv
...
+			if err := n.gethService(&les); err != nil {
+				log.Error("failed to get LES service")
+				return err
+			}
+
+			genesisHash := les.BlockChain().Genesis().Hash()
+			les1Topic := discv5.Topic("LES@" + common.Bytes2Hex(genesisHash.Bytes()[0:8]))
+			les2Topic := discv5.Topic("LES2@" + common.Bytes2Hex(genesisHash.Bytes()[0:8]))
+
+			n.config.RequireTopics[les1Topic] = params.Limits{Min: 1, Max: 2}
```
Comment (dshulyak) [inline node/status_node.go:130]: i noticed that it is better not to use different limits on end user devices. cause the longer we run discovery the more discovery protocol requests we will get which we will serve and in the end consume more mobile traffic.

File: peers/discv5.go [hunk 3]
```diff
@@ -59,7 +62,7 @@ func (d *DiscV5) Start() error {
 		return err
 	}
 	realaddr := conn.LocalAddr().(*net.UDPAddr)
-	ntab, err := discv5.ListenUDP(d.prv, conn, realaddr, "", nil)
+	ntab, err := discv5.ListenUDP(d.prv, conn, realaddr, "", d.version, nil)
```
Comment (dshulyak) [inline peers/discv5.go:65]: :+1:

File: peers/discv5.go [hunk 4]
```diff
@@ -95,3 +98,10 @@ func (d *DiscV5) Discover(topic string, period <-chan time.Duration, found chan<
 	d.net.SearchTopic(discv5.Topic(topic), period, found, lookup)
 	return nil
 }
+
+// InsertNodes inserts givens peers as nodes into Discovery V5 table.
+func (d *DiscV5) InsertNodes(topic discv5.Topic, nodes ...*discv5.Node) {
```
Comment (dshulyak) [inline peers/discv5.go:103]: just for info - in rendezvous node have to be registered every defined time period (i think 10s - 20s atm), after this time period it is purged from database. the reason for doing so is to guarantee better peer liveness.

**Review by dshulyak — COMMENTED** [review]
maybe it makes sense to run it as a separate executable? or at least separate entity? it seems that placing it under peer pool doesn't provide lots of benefits except that all discovery related things are under single entry point.

do you want to run it as a part of statusd or thats just because it was convenient?

File: node/status_node.go (hunk 2 above)
Comment (adambabik) [reply node/status_node.go:130]: As so you mean that probably `params.Limits{Min: 1, Max: 1}` or `params.Limits{Min: 2, Max: 2}` would be better?

**Comment by adambabik** [comment]
> maybe it makes sense to run it as a separate executable? or at least separate entity? it seems that placing it under peer pool doesn't provide lots of benefits except that all discovery related things are under single entry point.

> do you want to run it as a part of statusd or thats just because it was convenient?

I think we can definitely run it as a separate entity. Something like "bootstrap-proxy". My test scenario actually covers such a case when a node with proxy capability is not a peer of the final node. Distribution of peers with a particular topic is done via a bootstrap node. With a separate executable, we probably won't need any changes on the client side.

I think we can refactor `TopicPool` and `PeerPool` a bit. We already hacked it pretty bad for `cacheOnlyTopicPool`. I believe `TopicPool` can be split into smaller pieces and they can form a stuff like `TopicPoolWithCache` or `TopicPoolWithStop` etc.

File: node/status_node.go (hunk 2 above)
Comment (dshulyak) [reply node/status_node.go:130]: yes

**Comment by dshulyak** [comment]
@adambabik so what are we going to do now? if we want to use same thing for rendezvous - we will have to disable signature verification:

https://github.com/status-im/rendezvous/blob/master/server/server.go#L213

i guess it is not a big deal in our current situation

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/status-im/rendezvous">status-im/rendezvous</a></strong></div><div>Contribute to rendezvous development by creating an account on GitHub.</div></blockquote>

**Comment by adambabik** [comment]
@dshulyak can we introduce a notion of "proxied ENR records"? If one peer is not interested in that type of record, they can simply ignore them. Otherwise, the peer can treat them with a grain of salt. Without checking ENR signature, it would be quite easy to impersonate other peers in the network, right?

**Comment by dshulyak** [comment]
> Otherwise, the peer can treat them with a grain of salt. Without checking ENR signature, it would be quite easy to impersonate other peers in the network, right?

yes, initially i thought exactly about this problem, however it might be not a big issue.
what could be the worst thing if someone will do such thing? if node doesn't want to accept connections - it shouldn't accept them on the server side.
if the goal is to spam discovery - it can be done using properly signed records but non-dialable addresses.

**Comment by adambabik** [comment]
One case which might be affected is trusted nodes. Even if an ENR record is spoofed, it must contain correct `NodeID`, IP and port as all these three fragments are verified. Finally, the node will connect to a valid peer. 

However, it becomes dangerous if the ENR record contains more information than what is verified and this information affects peer selection. For example, the ENR can contain some information about current node load or free slots. In this way, the attacker can flood the network with invalid ENR records and convince nodes to redirect traffic to a particular peer.

**Comment by dshulyak** [comment]
yes indeed, that's definitely something that can be easily exploited. 

lets then extend ENR for proxied nodes. we can add one more field - proxied/proxy.
which will store public key of the proxied node. 

when client will fetch records it will have to check this field and get public key of the original node.
this way proxied nodes will be used only by clients that want to use them this way.

**Pull request closed without merging**